patches/server/1039-Add-more-native-map-methods-to-PDC.patch

```diff
@@ -0,0 +1,389 @@
++        Preconditions.checkArgument(value != null, "The provided value cannot be null");
++        
++        Z curvalue;
++        if (((curvalue = get(key, type)) != null) || has(key)) {
++            put(key, type, value);
++        }
++        return curvalue;
++    }
++
++    @Override
++    public <T, Z> Z merge(net.kyori.adventure.key.Key key, PersistentDataType<T, Z> type, Z value, java.util.function.BiFunction<? super Z, ? super Z, ? extends Z> remappingFunction) {
++        Preconditions.checkArgument(key != null, "The key cannot be null");
++        Preconditions.checkArgument(type != null, "The provided type cannot be null");
++        Preconditions.checkArgument(value != null, "The provided value cannot be null");
++        Preconditions.checkArgument(remappingFunction != null, "The provided remappingFunction cannot be null");
++        
++        Z oldValue = get(key, type);
++        Z newValue = (oldValue == null) ? value : remappingFunction.apply(oldValue, value);
++        if (newValue == null) {
++            remove(key);
++        } else {
++            put(key, type, newValue);
++        }
++        return newValue;
++    }
+ 
+     @Override
+     public boolean has(NamespacedKey key) {
++        return has((net.kyori.adventure.key.Key) key);
++    }
++    
++    @Override
++    public boolean has(net.kyori.adventure.key.Key key) {
+         Preconditions.checkArgument(key != null, "The provided key for the custom value was null");
+ 
+         return this.customDataTags.containsKey(key.toString());
+diff --git a/src/test/java/io/papermc/paper/persistence/PersistentDataContainerTest.java b/src/test/java/io/papermc/paper/persistence/PersistentDataContainerTest.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..e5a5dda6fb46aa58cd0e0b1235a0b121f7d0c66e
+--- /dev/null
++++ b/src/test/java/io/papermc/paper/persistence/PersistentDataContainerTest.java
+@@ -0,0 +1,156 @@
++package io.papermc.paper.persistence;
++
++import net.kyori.adventure.key.Key;
++import org.bukkit.NamespacedKey;
++import org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer;
++import org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry;
++import org.bukkit.persistence.PersistentDataType;
++import org.junit.jupiter.api.Test;
++
++import java.util.NoSuchElementException;
++import java.util.function.Supplier;
++
++import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
++import static org.junit.jupiter.api.Assertions.assertEquals;
++import static org.junit.jupiter.api.Assertions.assertNotEquals;
++import static org.junit.jupiter.api.Assertions.assertNull;
++import static org.junit.jupiter.api.Assertions.assertThrows;
++import static org.junit.jupiter.api.Assertions.assertTrue;
++import static org.junit.jupiter.api.Assertions.assertFalse;
++
++public class PersistentDataContainerTest {
++    
++    static private CraftPersistentDataContainer createContainer() {
```

Comment:
If you need to do something before each test you can use a @ BeforeEach method that would create an empty pdc for each test into a class field. Also you don't need to get a craft implementation, can use the api.